history.go

```diff
@@ -0,0 +1,75 @@
+package merkletree
+
+import (
+	"errors"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+)
+
+var (
+	ErrHistoryExisted = errors.New("[merkletree] trying to override existing history")
+	ErrBadEpoch       = errors.New("[merkletree] next epoch's STR has bad epoch ")
+)
+
+type History struct {
+	tree          *MerkleTree
+	currentSTR    *SignedTreeRoot
+	epochInterval int64
+}
+
+func NewHistory(m *MerkleTree, key crypto.SigningKey, startEp, epInterval int64) *History {
+	h := new(History)
+	h.tree = m
+	h.epochInterval = epInterval
+	h.currentSTR = newSTR(m, startEp, 0, make([]byte, crypto.HashSizeByte), key)
+	return h
+}
+
+func (h *History) UpdateHistory(m *MerkleTree, nextEp int64) error {
```

Comment:
Sorry, I might completely misunderstanding something here. 

The changes @arlolra suggesting is: insert new data to the tree, recompute the tree hash, clone the tree and generate new STR for new tree. 

So, at each epoch, the developer would do something like this:
```
func UpdateDirectory(h *History) {
    h.MerkleTree().Set(key, value)
    ....
    h.UpdateHistory()
}
```

Assume at epoch `t`, Alice send a registration request, then Alice's key would be inserted to the tree at epoch `t+1`. It means the Alice's key should not be found in the tree at epoch `t` but at epoch `t+1`, right?

But if we do as above, Alice's key would be found in the tree at epoch `t` (since we inserted to the tree before cloning it). 

Do I have any mistakes?